tree.go

```diff
@@ -148,65 +154,103 @@ func TreeFromChunks(chunks [][]byte) (*Node, error) {
 	for i, c := range chunks {
 		leaves[i] = NewNodeWithValue(c)
 	}
-	return TreeFromNodes(leaves)
+	return TreeFromNodes(leaves, numLeaves)
 }
 
 // TreeFromNodes constructs a tree from leaf nodes.
 // This is useful for merging subtrees.
-// The number of leaves should be a power of 2.
-func TreeFromNodes(leaves []*Node) (*Node, error) {
+// The limit should be a power of 2.
+// Adjacent sibling nodes will be filled with zero order hashes that have been precomputed based on the tree depth.
+func TreeFromNodes(leaves []*Node, limit int) (*Node, error) {
 	numLeaves := len(leaves)
 
-	if numLeaves == 1 {
+	depth := floorLog2(limit)
+	zeroOrderHashes := getZeroOrderHashes(depth)
+
+	// there are no leaves, return a zero order hash node
+	if numLeaves == 0 {
+		return NewEmptyNode(zeroOrderHashes[0]), nil
+	}
+
+	// now we know numLeaves are at least 1.
+
+	// if the max leaf limit is 1, return the one leaf we have
+	if limit == 1 {
 		return leaves[0], nil
 	}
-	if numLeaves == 2 {
+	// if the max leaf limit is 2
+	if limit == 2 {
+		// but we only have 1 leaf, add a zero order hash as the right node
+		if numLeaves == 1 {
+			return NewNodeWithLR(leaves[0], NewEmptyNode(zeroOrderHashes[1])), nil
+		}
+		// otherwise return the two nodes we have
 		return NewNodeWithLR(leaves[0], leaves[1]), nil
 	}
 
-	if !isPowerOfTwo(numLeaves) {
+	if !isPowerOfTwo(limit) {
 		return nil, errors.New("number of leaves should be a power of 2")
 	}
 
-	numNodes := numLeaves*2 - 1
-	nodes := make([]*Node, numNodes)
-	for i := numNodes; i > 0; i-- {
-		// Is a leaf
-		if i > numNodes-numLeaves {
-			nodes[i-1] = leaves[i-numLeaves]
-		} else {
-			// Is a branch node
-			nodes[i-1] = NewNodeWithLR(nodes[(i*2)-1], nodes[(i*2+1)-1])
+	leavesStart := powerTwo(depth)
+	leafIndex := numLeaves - 1
+
+	nodes := make(map[int]*Node)
```

Comment:
Changes `nodes` to a map instead of a slice so that we can lookup child nodes using the computed left and right index.